blockchain/utxocache.go

```diff
@@ -0,0 +1,812 @@
+			"remaining, %.2f MiB (%.2f%%))", flushedEntries, remainingEntries,
+			memUsageMiB, memUsagePercent)
+	}
+
+	return nil
+}
+
+// MaybeFlush conditionally flushes the cache to the database.
+//
+// If the maximum size of the cache has been reached, or if the periodic flush
+// interval has been reached, then a flush is required.  Additionally, a flush
+// can be forced by setting the force flush parameter.
+//
+// This function is safe for concurrent access.
+func (c *UtxoCache) MaybeFlush(bestHash *chainhash.Hash, bestHeight uint32,
+	forceFlush bool, logFlush bool) error {
+
+	c.cacheLock.Lock()
+	if forceFlush || c.shouldFlush(bestHash) {
+		err := c.flush(bestHash, bestHeight, logFlush)
+		c.cacheLock.Unlock()
+		return err
+	}
+
+	c.cacheLock.Unlock()
+	return nil
+}
+
+// InitUtxoCache initializes the utxo cache by ensuring that the utxo set is
+// caught up to the tip of the best chain.
+//
+// Since the cache is only flushed to the database periodically, the utxo set
+// may not be caught up to the tip of the best chain.  This function catches the
+// utxo set up by replaying all blocks from the block after the block that was
+// last flushed to the tip block through the cache.
+//
+// This function should only be called during initialization.
+func (b *BlockChain) InitUtxoCache(tip *blockNode) error {
+	log.Infof("UTXO cache initializing (max size: %d MiB)...",
+		b.utxoCache.maxSize/1024/1024)
+	defer log.Info("UTXO cache initialization completed")
+
+	// Fetch the utxo set state from the database.
+	var state *utxoSetState
+	err := b.db.View(func(dbTx database.Tx) error {
+		var err error
+		state, err = dbFetchUtxoSetState(dbTx)
+		return err
+	})
+	if err != nil {
+		return err
+	}
+
+	// If the state is nil, update the state to the tip.  This should only be the
+	// case when starting from a fresh database or a database that has not been
+	// run with the utxo cache yet.
+	if state == nil {
+		state = &utxoSetState{
+			lastFlushHeight: uint32(tip.height),
+			lastFlushHash:   tip.hash,
+		}
+		err := b.db.Update(func(dbTx database.Tx) error {
+			return dbPutUtxoSetState(dbTx, state)
+		})
+		if err != nil {
+			return err
+		}
+	}
+
+	// Set the last flush hash and the last eviction height from the saved state
+	// since that is where we are starting from.
+	b.utxoCache.lastFlushHash = state.lastFlushHash
+	b.utxoCache.lastEvictionHeight = state.lastFlushHeight
+
+	// If state is already caught up to the tip, return as there is nothing to do.
+	if state.lastFlushHash == tip.hash {
+		return nil
+	}
+
+	// Create a new view to use for disconnecting and reconnecting blocks.
+	view := NewUtxoViewpoint(b.utxoCache)
+	view.SetBestHash(&tip.hash)
+
+	// Iterate backward from the tip to the block after the one that was last
+	// flushed to get all of the blocks that need to be replayed through the
+	// cache.
+	replayNodes := make([]*blockNode, uint32(tip.height)-state.lastFlushHeight)
```

Comment:
While testing the reconciliation paths and taking a closer look, this does not seem to be handling the reorg case:

In other words:

```
A -> B  -> C  -> D -> E -> F <--- chain tip is now here
 \-> B' -> C' -> D' <--- cache is here, but there was a reog
```

For the cache to be correct, `D'`, `C'`, and `B'` need to be undone before replaying the attaches.

Reply:
For reference, I was able to hit this case on `simnet` with a slight bit of cheating to disable the automatic flush on shutdown so I didn't have to try to catch killing it at exactly the right moment, but it otherwise has been recovering nicely including physically removing the hardware in the middle of a sync.

Reply:
Good catch.  Thanks for pointing this out.

I took a look at handling this by recovering during initialization but ran into one issue in that the spend journal information for the side chain blocks will not be available.  Therefore, it is not simple to disconnect the side chain blocks before replaying the blocks to attach.

I made [an update to force flush](https://github.com/decred/dcrd/compare/82bc4a755fb6af64f4da03ed80f1d479b6da9c84..f4e2a54b2adceeffd2e5c33a948cce21bafd4a53) the cache when disconnecting blocks to avoid this scenario from occurring.  However, this still leaves open the possibility of a bad state if the database tx for disconnecting the block/spend journal is committed successfully, but then there is a failure before the utxo set is flushed.  It's also not trivial to combine those database transactions since that would then also require rolling back the view committing to the cache if the combined database transaction failed for any reason.

I'll probably need to either recreate the spend journal info or make updates to guarantee that the utxo cache stays in sync when disconnecting blocks.  I'll look into this further and see what works best.

Reply:
I think you can probably just move the code that removes the side chain block spend journal entries (`dbRemoveSpendJournalEntry`) to after the cache flush logic to ensure the entries are not removed while they are still potentially needed.

Reply:
I went with your suggestion to move the spend journal removal to after the cache flush logic.  As we discussed separately, the one caveat is that it is possible (though unlikely) that a block gets disconnected but then something errors out before removing the spend journal entry, which could leave a side block journal entry around indefinitely.  However, the only impact of this is using a bit of unnecessary disk space for an extremely rare case.

I reproduced the issue, and with this change plus adding the logic to disconnect blocks when necessary in the initialization, it is working properly now.  To make it easier to simulate the issue, I did the following:

- Comment out MaybeFlush so that it just returns nil (to avoid flushing in all cases)
- In disconnect block, return after disconnecting the block (but before flushing the cache or removing the spend journal).  This simulates some error occurring where the block was disconnected, but the cache was not flushed.
- Startup `dcrd` and invalidate some blocks / force a reorg.  At this point, the state of the flushed cache is on a side chain.
- Undo the above changes and restart `dcrd`.
- The utxo cache will need to disconnect (and potentially reconnect) blocks to catch up to the tip properly.

Reply:
I've reviewed the latest updated and confirm they work as expected and resolved the issue using a similar procedure to the one outlined here (along with some debug prints to confirm the expected reorg path was taken).  Looks good!